### DISCUSSION TITLE
Limiting case 2: only coupling, odd equilibrium

### DISCUSSION THREAD

**Discussion by jv426** [Part-II-Computational-Physics/.github#63] [General]
Hello!
I'm having some trouble with only coupling part of the exercise where we have to find how well our model fits to the analytical solution. Sometimes my code finds a really good fit, but every other time it finds an equilibrium at around +-0.25. I have been debugging the code for quite a few hours now and I can't seem to find the mistake. Could you please help me?

![Monte_Carlo_Equilibrium](https://user-images.githubusercontent.com/125775172/223612751-5b188fb9-ec13-47c1-a941-05612b21d6fa.png)

The relevant code is here:

def neighbouring_spins_sum(i,j,lattice,width):
    '''Sums the spins of all neighbours of the spin at [i,j].'''
    pos=neighbouring_sites(i,j,width)
    spins=[lattice[ii, jj] for ii, jj in pos]
    return sum(spins)
    
def compute_magnetisation(lattice):
    '''Computes the magnetisation of the lattice.'''
    N=len(lattice)
    N=N**2
    M=np.sum(lattice)/N
    return M

def compute_betaDeltaE(i,j,lattice,width,betaJ,betaMuH):
    '''Computes the energy difference between the old and new state if spin [i,j] would be flipped.'''
    spin=lattice[i,j]
    deltaE=2*betaJ*spin*neighbouring_spins_sum(i,j,lattice,width)+2*betaMuH*spin
    return deltaE

def attempt_spin_flip(lattice,width,betaJ,betaMuH):
    '''Applies the Metropolis-Hastings algorithm to try and flip a spin.'''
    i,j= np.random.randint(width, size=2)
    deltaE=compute_betaDeltaE(i,j,lattice,width,betaJ,betaMuH)
    if deltaE<=0 or np.random.random()<np.exp(-deltaE):
        lattice[i,j]=-lattice[i,j]
    return lattice

def evolve_and_compute_M(lattice,betaJ,betaMuH,avg_times):
    '''Evolves the lattice using the Metropolis-Hastings algorithm and returns the average magnetisation computed using different time steps.'''
    M=[]
    for t in range(avg_times[-1]+1):
        if t in avg_times:
            M.append(compute_magnetisation(lattice))
        attempt_spin_flip(lattice,len(lattice),betaJ,betaMuH)
    return np.mean(M)

betaJ_analytical = np.linspace(-5,5,100)
# Simuatle the system at different coupling strengths
def analyticalsol(beta):
    if beta<=0.5*np.log(1+np.sqrt(2)):
        return 0
    else:
        return (1-np.sinh(2*beta)**(-4))**(1/8)
    
M_analyticalpos=[]
M_analyticalneg=[]
avg_times =[100000, 110000, 120000, 130000, 140000] # Pick some suitable time points at which to sample your magnetisation. These values should be averaged to get a stable mean
lattice=init_lattice(50,0)
evolve=[]
for beta in betaJ_analytical:
    M_analyticalpos.append(analyticalsol(beta))
    M_analyticalneg.append(-analyticalsol(beta))
    evolve.append(evolve_and_compute_M(lattice, beta, 0, avg_times))

plt.plot(betaJ_analytical, M_analyticalpos, label="analytical positive")
plt.plot(betaJ_analytical, M_analyticalneg, label="analytical negative")
plt.plot(betaJ_analytical, evolve, label="Monte Carlo")
plt.xlabel ('Coupling strength betaJ')
plt.ylabel ('Magnetisation M')
plt.legend(loc='upper left')
plt.show()

**Comment by DLHvdH** [comment]
Check the code for computing the magnetisation again. A lattice with all spins up (+1) should give a magnetisation of 1 and vice-versa.

**Reply by jv426** [reply]
hmm, it prints the correct value for +1 and -1 lattices. Do you maybe have some other ideas?

**Reply by DLHvdH** [reply]
Just to be sure, `attempt_spin_flip(lattice,len(lattice),betaJ,betaMuH)` in `evolve_and_compute_M(lattice,betaJ,betaMuH,avg_times):` is outside the IF statement above?

I can't seem to spot your mistake. Have you tried starting with a +1 or -1 lattice for betaJ values above the critical point (see the hint)?

**Reply by jv426** [reply]
I think I got it working now, thanks!
